### PR TITLE
Feature/pay load reduction

### PR DESCRIPTION
**Pull request by etelmon**
### What does this do? 
This the phase 1 of sync lib being able to work with incremental changes to reduce network traffic.

- Add the incremental change update feature to sync (zerv. processUpdate)
- Add subscription fixes and large refactoring that were provided in release 1.2, but never pushed to develop
  - Performance improvements (log impact reduction)
  - Fixed publication filtering (Client does no longer receive all notifications)
  - Fixed publication formatting (to build the data transfer object at the publication level)
  - Increased reliability of notification removal (concurrent update and removal had no predicable result, now client will at least receive the update)
  - updated log lib

### Are There Any Performance Considerations?
Major increase during update.

### Additional Notes
there is a new version of zerv-ng-sync.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "zerv-sync",
-  "version": "1.1.3",
+  "version": "1.1.4",
```
Comment (Jacob-Friesen) [inline package.json]: The README.md needs to contain install instructions. I think only `npm install` is needed. But also mention the recommended version of Node.js and NPM to use there.

File: package.json [hunk 2]
```diff
@@ -13,24 +13,25 @@
     "url": "https://github.com/z-open/zerv-sync.git"
   },
   "scripts": {
-    "test": "./node_modules/jasmine/bin/jasmine.js"
+    "test": "./node_modules/jasmine/bin/jasmine.js",
+    "eslint": "./node_modules/eslint/bin/eslint.js --fix **/*.js"
```
Comment (Jacob-Friesen) [inline package.json]: This command `./node_modules/eslint/bin/eslint.js --fix **/*.js` gives the following errors:
```
/Users/jacobfriesen/Documents/work/zerv-sync/lib/lock.service.js
  11:5  error  'lockn' is assigned a value but never used  no-unused-vars

/Users/jacobfriesen/Documents/work/zerv-sync/lib/sync.helper.js
  4:7  error  'assert' is assigned a value but never used  no-unused-vars
  6:7  error  'logger' is assigned a value but never used  no-unused-vars

/Users/jacobfriesen/Documents/work/zerv-sync/lib/zerv-sync.js
  69:5  error  Parsing error: 'return' outside of function

✖ 4 problems (4 errors, 0 warnings)
```

File: lib/lock.service.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+ * Note:
+ * -----
+ * 1/ The proper way is to use db transaction.
+ *
+ * 2/ But if the process cannot be rollback. A simpler transaction scheme might be used to prevent access to some object.
+ * the functionality should use redis to publish the lock.
+ * Using the getLock would be needed to use in the entire application to decrease probability of data corruption. 
+ * lock should have a life span.
+ * 
+ * @param {*} id
```
Comment (Jacob-Friesen) [inline lib/lock.service.js]: You need to explain what id this is here.

File: lib/publication.js [hunk 4]
```diff
@@ -22,85 +22,126 @@ function Publication(name, fetchFn, dataNotifications, options) {
...
+    //         const r = this.dataNotifications[dataNotification].format(object, notificationType);
+    //         if (!r.then) {
+    //             return Promise.resolve(r);
+    //         }
+    //         return r;
+    //     } catch (err) {
+    //         logger.error(err);
+    //         return Promise.reject(err);
+    //     }
+    // };
```
Comment (Jacob-Friesen) [inline lib/publication.js]: These comments should be erased if they are no longer relevant or this needs to be an explanation as to why they are kept.

File: lib/subscription.js [hunk 5]
```diff
@@ -130,29 +136,17 @@ function Subscription(user, subscriptionId, publication, params) {
...
-                record = {
-                    id: getIdValue(record.id),
-                    revision: getRecordRevision(record),
-                    timestamp: {remove: new Date()},
-                    remove: new Date() // should remove in the zerv-ng-sync, and use the timestamp!!
-                };
-            }
+        _.forEach(notifications, (notification) => {
+            const record = notification.type === 'REMOVAL' ? buildLightRecordForRemoval(notification.record) : notification.record;
             changesToEmit |= addToQueue(record);
```
Comment (Jacob-Friesen) [inline lib/subscription.js]: Is this mean to be a bitwise or, it seems unlikely: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Assignment_Operators#Bitwise_OR_assignment

File: lib/sync.helper.js [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+                delete jsonObj[property];
+            } else {
+                mergeChanges(jsonObj[property], newValue);
+            }
+        } else {
+            jsonObj[property] = newValue;
+        }
+    });
+    return jsonObj;
+}
```
Comment (Jacob-Friesen) [inline lib/sync.helper.js]: Same comment as here. Due to the complexity of these functions, they need individual unit tests: https://github.com/z-open/zerv-ng-sync/pull/20#discussion_r305486829

File: package.json (hunk 2 above)
Comment (etelmon) [reply package.json]: Fixed

File: lib/sync.helper.js (hunk 6 above)
Comment (etelmon) [reply lib/sync.helper.js]: logic and unit tests moved to zjsonbin as it is shared between server and client

File: lib/lock.service.js (hunk 3 above)
Comment (etelmon) [reply lib/lock.service.js]: fixed

File: lib/publication.js (hunk 4 above)
Comment (etelmon) [reply lib/publication.js]: Removed.

File: lib/subscription.js (hunk 5 above)
Comment (etelmon) [reply lib/subscription.js]: Result would be the same here. but it is just boolean stuff. fixed for clarity

File: package.json (hunk 1 above)
Comment (etelmon) [reply package.json]: Added

File: lib/sync.helper.js (hunk 6 above)
Comment (Jacob-Friesen) [reply lib/sync.helper.js]: ok. As I said with the other PR, those still need to be improved: https://github.com/z-open/zjsonbin/blob/master/test/specs/zjson-bin.spec.js

**Pull request merged**